modules/contact/src/ContactAction.C

```diff
@@ -97,112 +108,296 @@ ContactAction::ContactAction(const InputParameters & params)
+  const std::string normal_lagrange_multiplier_name = action_name + "_normal_lm";
+  const std::string tangential_lagrange_multiplier_name = action_name + "_tangential_lm";
+
+  if (_current_task == "add_mesh_generator")
+  {
+    // Don't do mesh generators when recovering.
+    if (!(_app.isRecovering() && _app.isUltimateMaster()) && !_app.masterMesh())
     {
-      displacements.push_back(getParam<VariableName>("disp_y"));
-      if (isParamValid("disp_z"))
-        displacements.push_back(getParam<VariableName>("disp_z"));
-    }
+      const MeshGeneratorName master_name = master_subdomain_name + "_generator";
+      const MeshGeneratorName slave_name = slave_subdomain_name + "_generator";
 
-    mooseDeprecated("use the `displacements` parameter rather than the `disp_*` parameters (those "
-                    "will go away with the deprecation of the Solid Mechanics module).");
+      auto master_params = _factory.getValidParams("LowerDBlockFromSidesetGenerator");
+      auto slave_params = _factory.getValidParams("LowerDBlockFromSidesetGenerator");
+
+      master_params.set<MeshGeneratorName>("input") = _mesh_gen_name;
+      slave_params.set<MeshGeneratorName>("input") = master_name;
+
+      master_params.set<SubdomainName>("new_block_name") = master_subdomain_name;
+      slave_params.set<SubdomainName>("new_block_name") = slave_subdomain_name;
+
+      master_params.set<std::vector<BoundaryID>>("sidesets") = {_mesh->getBoundaryID(_master)};
+      slave_params.set<std::vector<BoundaryID>>("sidesets") = {_mesh->getBoundaryID(_slave)};
+
+      _app.addMeshGenerator("LowerDBlockFromSidesetGenerator", master_name, master_params);
+      _app.addMeshGenerator("LowerDBlockFromSidesetGenerator", slave_name, slave_params);
+    }
   }
 
-  // Determine number of dimensions
-  const unsigned int ndisp = displacements.size();
+  if (_current_task == "add_variable")
+  {
+    // Add the lagrange multiplier on the slave subdomain.
+    const auto addLagrangeMultiplier =
+        [this, &slave_subdomain_name, &displacements](const std::string & variable_name,
+                                                      const int codimension) //
+    {
+      InputParameters params = _factory.getValidParams("MooseVariableBase");
 
-  // convert vector of VariableName to vector of VariableName
-  std::vector<VariableName> coupled_displacements(ndisp);
-  for (unsigned int i = 0; i < ndisp; ++i)
-    coupled_displacements[i] = displacements[i];
+      mooseAssert(_problem->systemBaseNonlinear().hasVariable(displacements[0]),
+                  "Displacement variable is missing");
+      const auto primal_type =
+          _problem->systemBaseNonlinear().system().variable_type(displacements[0]);
+      const int lm_order = primal_type.order.get_order() - codimension;
 
-  if (_current_task == "add_dirac_kernel")
+      if (primal_type.family == MONOMIAL || (primal_type.family == LAGRANGE && lm_order < 1))
```

Comment:
Good catch. I added a clamp to zero to allow mortar with 0th order elements since a 0th order element should work (shouldn't it?). But if you prefer an error instead, I will fix.

Reply:
Umm, there are two different concepts from libMesh. The geometric element will have a minimum of first order...e.g. a first order Lagrange polynomial can map between reference and physical coordinates. If an element is second order, you would need a second order Lagrange polynomial to do the mapping, etc.

However, the basis functions associated with approximation solutions can be all sorts of things. Lagrange basis functions have a minimum of first order. Monomial basis functions can be zeroth order, e.g. constant across an element.

Reply:
Yes this clamp looks fine